### ISSUE TITLE
TensorFlow backward_passes_per_step>1 not work

### ISSUE THREAD

**Issue by PANXiao1994** [horovod/horovod#1676]
**Environment:**
1. Framework: TensorFlow
2. Framework version: 1.15.0
3. Horovod version: 0.19.0
4. MPI version: Open MPI: 2.0.2
5. CUDA version: CUDA Version 10.0.130
6. NCCL version: libnccl-dev=2.4.8-1
7. Python version: python 3.5
8. OS and version: Debian GNU/Linux,   9 (stretch)
9. GCC version: gcc version 6.3.0 20170516 (Debian 6.3.0-18+deb9u1)

**Bug report:**
Please describe errorneous behavior you're observing and steps to reproduce it.
- This is a continuation of https://github.com/horovod/horovod/issues/1667

In order to figure out the cause of the degradation, I read the source code of the class [_DistributedAdasumOptimizer](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/__init__.py#L313), which will be called when `backward_passes_per_step>1`.  **I just wonder whether there is any issue about synchronization** [here](https://github.com/horovod/horovod/blob/master/horovod/tensorflow/__init__.py#L337)? 

Actually, after reading the source code, I realized that this is the idea is to do one allreduce operation every `backward_passes_per_step` steps, instead of do `backward_passes_per_step` backward passes per step... ... So should it be`actual global step`=`global_step`/`backward_passes_per_step`?

```
        def _apply_shared(self, var, get_update_op):
            start_slot = self._get_or_make_slot(var, "delta_start")

            # initialize start on the first step
            assign_op = tf.cond(self._is_first_step, 
                lambda: start_slot.assign(var, use_locking=self.use_locking).op, 
                tf.no_op)
            
            with tf.control_dependencies([assign_op]):
                update_op = get_update_op()
                with tf.control_dependencies([update_op]):
                    def update():
                        # delta = var - start
                        local_delta = var.assign_sub(start_slot, use_locking=self.use_locking) # reuse var's memory
                        # delta = allreduce (delta)
                        global_delta = allreduce(local_delta,
                                                 device_dense=self._device_dense,
                                                 device_sparse=self._device_sparse,
                                                 compression=self._compression,
                                                 op=Adasum)
                        # start = start + delta
                        new_start = start_slot.assign_add(global_delta, use_locking=self.use_locking)
                        # var = start
                        return var.assign(new_start, use_locking=self.use_locking).op
                    
                    # if its a communication step, then apply logic above
                    # if its not a communication step then just have the underlying
                    # optimizer update the model parameters according to its logic
                    return tf.cond(self._is_comm_step, update, tf.no_op)
```

In my log, there is a piece of evidence that this is a synchronization issue:
```
I0119 18:23:14.160345 140281835146048 hooks.py:350] (2020-01-19 18:23:14.160261)Update 1600 	 TrainingLoss=553.723816   UD 0.131382 secs/step
INFO:tensorflow:(2020-01-19 18:23:14.740350)Update 2000 	 TrainingLoss=332.019073   UD 0.114731 secs/step
I0119 18:23:14.740418 140516833930048 hooks.py:350] (2020-01-19 18:23:14.740350)Update 2000 	 TrainingLoss=332.019073   UD 0.114731 secs/step
INFO:tensorflow:(2020-01-19 18:23:23.392424)Update 1600 	 TrainingLoss=331.872070   UD 0.134689 secs/step
I0119 18:23:23.392499 139727902747456 hooks.py:350] (2020-01-19 18:23:23.392424)Update 1600 	 TrainingLoss=331.872070   UD 0.134689 secs/step
INFO:tensorflow:(2020-01-19 18:23:27.618606)Update 1600 	 TrainingLoss=182.322266   UD 0.113972 secs/step
I0119 18:23:27.618671 139642867927872 hooks.py:350] (2020-01-19 18:23:27.618606)Update 1600 	 TrainingLoss=182.322266   UD 0.113972 secs/step
INFO:tensorflow:Evaluating DEVSET: DevLoss=178.572727  GlobalStep=1600
I0119 18:23:28.714222 139642867927872 text_metrics_spec.py:265] Evaluating DEVSET: DevLoss=178.572727  GlobalStep=1600
```

Where one of the gpu reached global step 2000 before others... 

Another evidence is that the time for each step to complete remains the same regardless of `hvd.size()`, which is counter-intuition....

**Comment by tgaddair** [comment]
Hey @Tixxx, can you help with looking into this issue?

**Comment by Tixxx** [comment]
Hi @PANXiao1994, yes the purpose of using `backward_passes_per_step` is to reduce the number of allreduce communications by accumulating gradients computed from `backward_passes_per_step` number of back-props per GPU, so we don't actually need to consider global step in this case.  Users will also need to call optimizer.step() every `backward_passes_per_step` number of minibatches to make this work.
As for the synchronization issue, I think it's caused by a bug in our code. It's not considering the case where all ranks won't have the even amount of data so some ranks might finish processing all their data before `backward_passes_per_step` is reached. Just to confirm this, in your example, do all the ranks eventually reach 2000 steps?

> Another evidence is that the time for each step to complete remains the same regardless of hvd.size(), which is counter-intuition....

Does your step mean each optimizer.step()? If so, the time will be dependent on `backward_passes_per_step`. With a large number of gradient accumulations, the computation time takes over so change in communication time(due to changing hvd.size()) is really negligible.
